### DISCUSSION TITLE
Understanding of git

### DISCUSSION THREAD

**Discussion by nimrossum** [git-truck/git-truck#46] [Research]
## [Git - Packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles)
When a repository gets large, Git optimize it by removing redundancy. We need to be able to handle Packfiles. We should consider using `git cat-file`, since it handles all these things already and probably better than we can do by ourselves.

**Comment by emiljapelt** [comment]
# Manual handling of git objects
There are two cases that need to be handled. "Raw" git objects, and packed git objects.
## "These objects are f**king raw!!!!"
There are four types of git objects, Commit, Tree, Blob and Tag. Tag is currently disregarded. 
### Commit
These contain compressed plaintext, so they can easily be decompressed and analyzed with zlib.unzip/zlib.decompress.
### Tree
These contain a combination of, metadata in plaintext, and object hashes in binary. This means that it is entirely simple to get the hash 
of a contained objects of the tree. To do this, the entire object must be decompressed via zlib, and then reformatted. The reformatting, is quite simply just to get the hashes as plaintext, and can be done be looping over the array of characters, and switching the loop behaviour when encountering the null character. 
### Blob
These contain compressed plaintext, so they can easily be decompressed and analyzed with zlib.unzip/zlib.decompress.
## [Packs](https://git-scm.com/docs/pack-format)
A pack consists of two files, an index (pack-\*.idx) and the contents (pack-\*.pack). The index file allows us to quickly gather the information we need, to efficiently find the contents of the objects stored in the pack. The exist two versions of the index. Quite a few of the data points stored in these files, are 4B number in the network byte order format.
### Index v1
Starts with a first-level fanout-table. Simply put this table is 256 4B numbers. The *n*'th number represents the amount of objects in the pack, which hashes start with a value less than or equal to *n*. The total number of objects in the pack is therefore obtainable from the last entry in the table.
Following this table are 24B entries for each object in the pack. Each entry starts with a 4B number, which is the offset required to reach the objects data in the .pack file, followed by the name of the object.
Finally are two checksums, one for the .pack file, and one the .idx file.
### Index v2
This index starts with a 4B magic number (\377tOc), which would not make sense as the 0'th entry in the fanout table. Followed by a 4B version number, which is 2. Then comes the fanout table, which is exactly like the one in v1.
Then comes a table of sorted object names, with no additional information.
Then another table with 4B CRC32 vales for each object, what ever that is (I dont think these are relevant for our purposes).
Then comes yet another table of 4B numbers. Each number is 31-bit, if MSB is 0, then then number is an offset into the .pack file, else the number points to an entry in the following table.
Which is a table of 8B numbers, all of which are offsets into the .pack file.
The file then contains the same checksums, that v1 contain.
### Content
This file contains a 4B signature, a 4B version, a 4B number of objects in the pack. Following this comes the actual objects. Each object entry starts with an object type and lenght, and then the compressed contents. Packs can have two new objects, offset deltas and reference deltas, which are instructions that should be applied on the base object, to obtain another version of it. The compresses content might behave exactly like the raw objects, i.e. obtainable by decompression, not sure.
